### DISCUSSION TITLE
How to send headers in deno?

### DISCUSSION THREAD

**Discussion by scientific-dev** [denoland/deno#9891] [Q&A] [answered]
I am working with an socket api where i need to send headers with the socket connections. However in nodejs it allows me to input headers in the second param. But as deno uses the browser like `WebSocket` api. I can't send headers with it. Any way so how can i send it with deno?

**Comment by crowlKats** [comment]
You can't set headers for the WebSocket API, as the spec doesnt allow such. See https://github.com/whatwg/html/issues/3062

**Reply by justinmchase** [reply]
Is this really the position of Deno? Many websocket API's rely on headers for authentication. The lower level websocket protocol already relies on headers to upgrade and handshake, and _many_ other implementations require more headers for authentication.

So while it may not be part of the WebSocket spec itself it is required to be even usable.

**Reply by caspervonb** [reply]
> Is this really the position of Deno? Many websocket API's rely on headers for authentication. The lower level websocket protocol already relies on headers to upgrade and handshake, and _many_ other implementations require more headers for authentication.
> 
> So while it may not be part of the WebSocket spec itself it is required to be even usable.

Do you have some examples of APIs in the wild that do this?

**Reply by crowlKats** [reply]
Deno implements web specifications as they are described to ensure compatibility and consistent behaviour with the web. Websockets are a large part of various browser-based services across the globe, and they all seem fine without supporting headers. 
Now, I do agree that setting headers would be quite beneficial, but we can't just go and break standards. Only options we have is to either wait for the specification to adapt or find some solution to work around this without going against the spec.

**Reply by justinmchase** [reply]
> Do you have some examples of APIs in the wild that do this?

The kubernetes API and GraphQL apis, or really any server that wants to check authentication before upgrading the connection.

**Reply by justinmchase** [reply]
> Deno implements web specifications as they are described... we have is to either wait for the specification to adapt or find some solution to work around this without going against the spec..

Maybe I'm missing something but it _is_ already in the specification. The ability to add additional headers _is_ already a part of the WebSocket protocol. This is why all of the existing implementations support it.

[source](https://datatracker.ietf.org/doc/html/rfc6455) 

[Section 1.2](https://datatracker.ietf.org/doc/html/rfc6455#section-1.2) Protocol Overview

> An unordered set of header fields comes after the leading line in
   both cases.  The meaning of these header fields is specified in
   Section 4 of this document.  Additional header fields may also be
   present, such as cookies [RFC6265].  The format and parsing of
   headers is as defined in [RFC2616].

For emphasis it says right in the protocol overview:

> ### Additional header fields may also be present, such as cookies.

Am I'm misreading or looking at the wrong RFC?

**Reply by crowlKats** [reply]
I am talking about the spec for the actual API, not the protocol: https://html.spec.whatwg.org/multipage/web-sockets.html

**Reply by justinmchase** [reply]
![giphy](https://user-images.githubusercontent.com/10974/125836183-a9cd9023-2149-46e9-9418-d7898b91d91e.gif)

Its the most html thing ever for the "standard" to differ from the "specification". But thanks for the clarification, I also see the epic discussion thread in their github repo which appears to be fairly recalcitrant about adding for reasons I don't quite understand.

The problem I'm encountering is that it seems like the kubernetes API requires authorization headers for websocket connections and the nodejs library supports it but Deno doesn't. So I realize Deno is trying to be pure here but it may end up being a casualty to practicality in the end.

**Reply by justinmchase** [reply]
I think my final argument here is going to be that the limitation of not being able to add headers is specifically a limitation of the browser and does not make sense in a server environment and the options and headers should be added here.

Now I know Deno is trying to (rightly) conform many of its apis to established web standards and apis and I think thats a good thing but it is also probably reasonable to deviate slightly from those standards when it makes sense and this is probably one of those cases.

Maybe it would make sense for deno to have an additional flag such as `--allow-server` which would enable portions of its api that would otherwise be restricted when running in a browser? I don't know just a quick thought.

**Reply by lucacasonato** [reply]
The main issue is that API wise we don't have a logical extension point for this right now. The WebSocket constructor does not take an options bag, so we can't easily add more options. I do see why this is a problem though.

The new WebSocketStream constructor does take an options bag, so we can consider taking a new option for headers there.

Unrelated: is there documentation on the WebSocket support on the k8s API?

**Reply by justinmchase** [reply]
The documentation is pretty sparse I've been having to dig through code. Let me know if you find it :)

**Reply by justinmchase** [reply]
Related To https://github.com/cloudydeno/deno-kubernetes_client/issues/5

**Reply by justinmchase** [reply]
Hey guess what! I poked the bear yesterday and it said we were free to make our own WebSocket standard for non-browser environments! 🤣 

https://github.com/whatwg/html/issues/3062#issuecomment-881056952

**Reply by justinmchase** [reply]
Also here are some relevant links in code I found yesterday while investigating this:

Here is a link to the signature of the main WebSocket library in nodejs, which does include support for headers:
https://github.com/websockets/ws/blob/HEAD/doc/ws.md#new-websocketaddress-protocols-options

Here is the starting point in kubernetes codebase:
https://github.com/kubernetes/kubectl/blob/master/pkg/cmd/exec/exec.go#L366

Here is the official nodejs kubernetes client exec api, which uses `ws`:
https://github.com/kubernetes-client/javascript/blob/master/src/exec.ts#L53

Here it is applying the authorization header to the options object:
https://github.com/kubernetes-client/javascript/blob/master/src/config.ts#L435

Here it creates the WebSocket with those options:
https://github.com/kubernetes-client/javascript/blob/master/src/web-socket-handler.ts#L174

**Reply by kitsonk** [reply]
Sending headers over a web socket is, well, just using the socket to `.send()`. This is as simple has having a utility class in `std` which accepts a web browser standard `WebSocket` and sends headers over it.  There is no need to re-implement, modify or change the web standard.  It might make sense to continue to have it as a "sidecar", something like this:

```ts
interface SockeetHeadersOptions {
  headers?: Headers;
  autoSend?: boolean;
}

export class SocketHeaders {
  constructor(websocket: WebSocket, options: SocketHeadersOptions);
  get(key: string): string | undefined;
  set(key: string, value: string): void;
  send(): Promise<void>;
}
```

Deviating from the web standards should only be done when there is no other realistic alternative, because when/if we do it, it decreases portability of code. Using this sort of "side car" class increases portability because this code can likely be used on other platforms.

**Reply by crowlKats** [reply]
@kitsonk that doesn't make much sense; this would assume that the server would need to interpret websocket messages as headers themselves. What is asked here is to set additional headers during the initial handshake

**Reply by justinmchase** [reply]
@kitsonk I think that's partially right, the websocket standards maintainers certainly think that all communication should happen that way but the send happens after the handshake I believe and in this case the server wants to observe the headers before upgrading the connection and doing the handshake.

**Comment by scientific-dev** [comment]
I have viewed all of those and finally ended up here. So is there any other way?

**Comment by scientific-dev** [comment, marked as answer]
Well i managed making a ws module which supports custom headers...
https://github.com/scientific-dev/custom-socket
